### PR TITLE
Ubernetes-Lite GCE: Label volumes with zone information

### PR DESCRIPTION
**Pull request by justinsb**
I haven't yet had time to test this, but wanted to share with you asap @quinton-hoole 

This completes Ubernetes-Lite support for GCE, by labeling volumes with zones.  It also enforces non-ambiguousness when referring to volumes purely by name, which ends up having some very nice properties: from commit 8dbb5c90b1da68c6ed01c770beeceeeedd973907

```
    Ubernetes-Lite: Error if a PD name is ambiguous

    We don't cope well if a PD is in multiple zones, but this is actually
    fairly easy to detect.  This is probably justified purely on the basis
    that we never want to delete the wrong volume (DeleteDisk), but also
    because this means that we now warn on creation if a disk is in multiple
    zones (with the labeling admission controller).

    This also means that with the scheduling predicate in place, that many
    of our volume problems "go away" in practice: you still can't create or
    delete a volume when it is ambiguous, but thereafter the volume will be
    labeled with the zone, that will match it only to nodes with the same
    zone, and then we query for the volume in that zone when we
    attach/detach it.
```

### CODE REVIEW

**Comment by quinton-hoole** [comment]
@justinsb Thanks!  I'll do an initial review now.

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -1644,6 +1645,30 @@ func (gce *GCECloud) DeleteDisk(diskToDelete string) error {
 	return gce.waitForZoneOp(deleteOp, disk.Zone)
 }
 
+func (gce *GCECloud) GetPDLabels(name string) (map[string]string, error) {
```
Comment (quinton-hoole) [inline pkg/cloudprovider/providers/gce/gce.go]: nit: Worth adding a comment here describing what this does, as it's not entirely clear from the signature. In particular it does not return all of the labels currently applied to the PD.  It returns the zone and region labels that we want to apply to the PD.  Also, the name param is the name of the PD, not the name of the label, or of anything else.

File: plugin/pkg/admission/persistentvolume/label/admission.go [hunk 2]
```diff
@@ -40,8 +41,9 @@ var _ = admission.Interface(&persistentVolumeLabel{})
 type persistentVolumeLabel struct {
 	*admission.Handler
 
-	mutex      sync.Mutex
-	ebsVolumes aws.Volumes
+	mutex            sync.Mutex
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
```
Comment (quinton-hoole) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: Can we avoid putting provider-specific fields in this provider-independent library?

Comment (justinsb) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: I don't see how - suggestions welcome.

I wanted to implement a Volumes interface, and you can see from the AWS implementation that I did.  It was originally intended to be cross-cloud provider, but there were a number of objections raised so I moved Volumes into the aws package.  The primary objection was (I believe) that a cloud could implement different types of volumes (e.g. even AWS has NFS & EBS now).

And volume.Spec isn't provider independent either.

One option would be to define aws.Volumes in cloudprovider, but have it take a PersistentVolumeSource.  But that isn't quite right either, and would be a fairly large refactor.

In short, I'd love to, but I don't see how to do it in an acceptable way.

Comment (quinton-hoole) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: Fair enough.  I'm happy for us to perhaps try to take on a refactor in a separate PR, after v 1.2.

**Comment by justinsb** [comment]
@k8s-bot test this

Looks like a flake?

**Comment by quinton-hoole** [comment]
@justinsb As discussed offline, I'm not terribly happy about how plugin/pkg/admission/persistentvolume/label/admission.go is morphing from what was presumably intended to be a provider-agnostic generalization into an ugly amalgamation of provider-specific code.  But I understand your point that much of this stems from previous decisions, and is not easy to fix in this PR. I'll open a separate issue to track the refactoring issue required to address that.  

Other than that, this PR LGTM!  Thanks!

**Comment by justinsb** [comment]
@k8s-bot unit test test

**Comment by justinsb** [comment]
@k8s-bot unit test this

**Comment by quinton-hoole** [comment]
Merging manually to free up some merge queue bandwidth.

**Comment by lavalamp** [comment]
@quinton-hoole Pinging @k8s-oncall for manual merges is probably a better way to go. :)

File: plugin/pkg/admission/persistentvolume/label/admission.go [hunk 3]
```diff
@@ -75,6 +77,13 @@ func (l *persistentVolumeLabel) Admit(a admission.Attributes) (err error) {
 		}
 		volumeLabels = labels
 	}
+	if volume.Spec.GCEPersistentDisk != nil {
```
Comment (thockin) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: @markturansky @saad-ali This should be part of volume plugins, I think.  It should NOT be an open-coded list of cases.  I'll file a new issue.

File: plugin/pkg/admission/persistentvolume/label/admission.go (hunk 2 above)
Comment (markturansky) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: If volumes are plugins and Kubelet (and others) don't know what plugins exist, anything that escapes those packages is leaky.

Would we instead want to give the same list of plugins to the API server that we give to Kubelet and ControllerManager?

The API server would do the same iteration and CanSupport on each plugin, find the appropriate one, and call Validate(pv) and Admit(pv).

Each volume would know how to validate itself in the API.  Missing plugins would cause a real API error.  Any business logic regarding labelling and so forth is also internalized to the plugin.

@thockin thoughts?

File: plugin/pkg/admission/persistentvolume/label/admission.go [hunk 4]
```diff
@@ -75,6 +77,13 @@ func (l *persistentVolumeLabel) Admit(a admission.Attributes) (err error) {
...
+		labels, err := l.findGCEPDLabels(volume)
+		if err != nil {
+			return admission.NewForbidden(a, fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err))
```
Comment (markturansky) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: This broke provisioning (because we don't have an e2e test that would have prevented this change).

The provisioner creates a dummy PV in response to a claim that requires provisioning.  The "dummy" ID on the struct was required to pass validation. A controller reconciles this PV with the infrastructure and creates a resource asynchronously.

There is no "dummy" PD in the provider yet, so we can't look it up for labels.  This admission controller makes it required that the PD exist.  It could be, too, that making a placeholder PV is wrong.  I am open to suggestions.

File: plugin/pkg/admission/persistentvolume/label/admission.go (hunk 3 above)
Comment (justinsb) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: Yes - this came up in review a few lines up ^^^.  I agree with the criticism, but it isn't clear to me how best to fix it.

File: plugin/pkg/admission/persistentvolume/label/admission.go (hunk 4 above)
Comment (justinsb) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: I opened #20564 to track the lack of e2e coverage and #20955 to track the AWS problem, but I didn't realize this had broken it on GCE also.  Sorry!

For AWS, I was thinking of creating a volume with a well-known dummy name, which would then be ignored.  But that is pretty evil...

Comment (markturansky) [inline plugin/pkg/admission/persistentvolume/label/admission.go]: we can put the "dummy" constant in pkg/volume (as opposed to a plugin) and use that in admission and the plugins themselves.

it is leaky as hell but is far less work than adding plugins and a framework to the API server.  Should we consider it as a short term fix until we come up with a better way?
